### ISSUE TITLE
Latest version of libobjc2 calculates wrong ivar offsets for embeded structures

### ISSUE THREAD

**Issue by minusbat** [gnustep/libobjc2#27]
I have an object which contains a 'struct stat' as an instance variable. The latest version of the runtime is allocating the correct amount of space for it (120 bytes on FreeBSD) but is setting the base address sixteen bytes hugher than it shiuld. Thus using the structure overflows into the following instance variables and corrupts them.

This only happens with the non-fragile ABI. Using the fragile ABI works fine.

Example code below - this prints out he addresses of the object, the istance varaoibles and the size of struct stat. Its easy to see that it is getting it wrong. I havent been able to identify which commit causes the issue yet, but I notice there have been several chnages to ivar layout from the comments.

File attached - complie with: cc -I/usr/local/include -L/usr/local/lib test.m -lobjc

[test.txt](https://github.com/gnustep/libobjc2/files/646557/test.txt)

**Comment by davidchisnall** [comment]
Curses, I thought I'd fixed all of the ivar-offset bugs.  Correctly aligning vector ivars seems to have a lot of fallout...

**Comment by minusbat** [comment]
That works beautifully, thanks! Out of interest, why move it up instead of down ? I was expecting it to go lower in memory so it was tight packed against the preceeding pointer value.

**Comment by davidchisnall** [comment]
We're moving ivars to ensure that they have enough alignment for their type.  Unfortunately, we're having to work this out from (sometimes) inadequate information as with the current ABI the compiler isn't providing the alignment (I'm working on this separately).  In the common case, we can't move things backwards because it will overlap the previous ivar.

**Comment by minusbat** [comment]
Ok,makes sense.Anyway, thanks for the quick fix again.

We are currently working on mving from BSD 10 to 11, so I may have some more issues for you over the comming days. Will try and idenfiy and fix them before they get this far though.

**Comment by davidchisnall** [comment]
What are you working on?

**Comment by minusbat** [comment]
Heh, same thing I have been for the last fifteen years :-) I think we met once in a pub at a GNUstep meetup at Liverpool Street a long long time ago ? Back in 2000 I started writing some middleware under Nt4 as a contract job. Being primarily a NeXTStep guy I did it in objective C on BSD (cross compiled staticly to Cygwin)  and built a little collection of basic objects to work with (String, Array, Dictionary, AutoreleasePool etc). GNUstep wasnt in a suitable place to be used at that point or I would have done.

So the two week contract job turned into an actual career, and I've carried that code with me the whole way. I finally converted it away from GCC to Clang about a year or so ago, moved to your runtime, and added support for some of the ObjC- 2.0 features.

I might be the last person in the world with their won set of base classes, but we've got 300k lines of code written in this, so I take a keen interest in runtimes and compilers, as the fate of the entire business depends on it :-)
